samza-log4j/src/main/java/org/apache/samza/logging/log4j/StreamAppender.java

```diff
@@ -110,16 +125,41 @@ public void append(LoggingEvent event) {
             log.trace("Waiting for the JobCoordinator to be instantiated...");
           }
         } else {
-          OutgoingMessageEnvelope outgoingMessageEnvelope =
-              new OutgoingMessageEnvelope(systemStream, key.getBytes("UTF-8"), serde.toBytes(subLog(event)));
-          systemProducer.send(SOURCE, outgoingMessageEnvelope);
+          // Serialize the event before adding to the queue to leverage the caller thread
+          // and ensure that the queueToProducerTransferThread can keep up.
+          if (!logQueue.offer(serde.toBytes(subLog(event)), queueTimeoutS, TimeUnit.SECONDS)) {
+            // Do NOT retry adding to the queue. Dropping the event allows us to alleviate the unlikely
+            // possibility of a deadlock, which can arise due to a circular dependency between the SystemProducer
+            // which is used for StreamAppender and the log, which uses StreamAppender. Any locks held in the callstack
+            // of those two code paths can cause a deadlock. Dropping the event allows us to proceed.
+
+            // Scenario:
+            // T1: holds L1 and is waiting for L2
+            // T2: holds L2 and is waiting to produce to BQ1 which is drained by T3 (SystemProducer) which is waiting for L1
+
+            // This has happened due to locks in Kafka and log4j (see SAMZA-1537), which are both out of our control,
+            // so dropping events in the StreamAppender is our best recourse.
+
+            // Drain the queue instead of dropping one message just to reduce the frequency of warn logs above.
+            int messagesDropped = logQueue.drainTo(new ArrayList<>()) + 1; // +1 because of the current log event
+            log.warn(String.format("Exceeded timeout %ss while trying to log to %s. Dropping %d log messages.",
+                queueTimeoutS,
+                systemStream.toString(),
+                messagesDropped));
+
+            // Emit a metric which can be monitored to ensure it doesn't happen often.
+            metrics.logMessagesDropped.inc(messagesDropped);
+          }
+          metrics.bufferFillPct.set(Math.round(100 * logQueue.size() / DEFAULT_QUEUE_SIZE));
         }
       } catch (Exception e) {
         System.err.println("[StreamAppender] Error sending log message:");
         e.printStackTrace();
       } finally {
         recursiveCall.set(false);
       }
+    } else if (metrics != null) {
```

Comment:
Yes, because metrics are initialized in setupSystem, which may not have been called for this code path, but has definitely been called for the others.